rfcs/lower-bounds-calculation.md

```diff
@@ -0,0 +1,215 @@
+# Lower Bounds Calculation
+
+## Summary
+
+We propose adapting lower bounds calculation from Pierce's Local Type Inference paper into the Luau type inference algorithm.
+
+https://www.cis.upenn.edu/~bcpierce/papers/lti-toplas.pdf
+
+## Motivation
+
+There are a number of important scenarios that occur where Luau cannot infer a sensible type without annotations.
+
+Many of these revolve around type variables that occur in contravariant positions.
+
+### Function Return Types
+
+A very common thing to write in Luau is a function to try to find something in some data structure.  These functions habitually return the relevant datum when it is successfully found, or `nil` in the case that it cannot.  For instance:
+
+```lua
+-- A.lua
+function find_first_if(vec, f)
+    for i, e in ipairs(vec) do
+        if f(e) then
+            return i
+        end
+    end
+
+    return nil
+end
+```
+
+This function has two `return` statements: One returns `number` and the other `nil`.  Today, Luau flags this as an error.  We ask authors to add a return annotation to make this error go away.
+
+We would like to automatically infer `find_first_if : <T>({T}, (T) -> boolean) -> number?`.
+
+Higher order functions also present a similar problem.
+
+```lua
+-- B.lua
+function foo(f)
+    f(5)
+    f("string")
+end
+```
+
+There is nothing wrong with the implementation of `foo` here, but Luau fails to typecheck it all the same because `f` is used in an inconsistent way.  This too can be worked around by introducing a type annotation for `f`.
+
+We would like to infer `f : <T...>((number) -> T... & (string) -> T...) -> ()`
```

Comment:
Should we infer an intersection here, or a union argument type? `<T...>(number|string) -> T...`? (Tricky without type pack unions.)

Reply:
Do you want to infer a union type in specific cases when it's possible?
Since it's sound only if a single parameter has different types in the intersection.

Reply:
We could maybe do this someday, but, as you say, type pack unions are a big feature all on their own.